protocol/linkpreview/linkpreview.go

```diff
@@ -241,13 +247,104 @@ func (u OpenGraphUnfurler) unfurl() (common.LinkPreview, error) {
 	return preview, nil
 }
 
+type ImageUnfurler struct {
+	url        *neturl.URL
+	logger     *zap.Logger
+	httpClient http.Client
+}
+
+func compressImage(imgBytes []byte) ([]byte, error) {
+	smallest := imgBytes
+
+	img, err := images.DecodeImageData(imgBytes, bytes.NewReader(imgBytes))
+	if err != nil {
+		return nil, err
+	}
+
+	compressed := bytes.NewBuffer([]byte{})
+	err = images.CompressToFileLimits(compressed, img, images.DefaultBounds)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(compressed.Bytes()) < len(smallest) {
+		smallest = compressed.Bytes()
+	}
+
+	if len(smallest) > maxImageSize {
+		return nil, errors.New("image too large")
+	}
+
+	return smallest, nil
+}
+
+func (u ImageUnfurler) unfurl() (common.LinkPreview, error) {
+	preview := newDefaultLinkPreview(u.url)
+	preview.Type = protobuf.UnfurledLink_IMAGE
+
+	headers := map[string]string{"user-agent": headerUserAgent}
+	imgBytes, err := fetchBody(u.logger, u.httpClient, u.url.String(), headers)
+	if err != nil {
+		return preview, err
+	}
+
+	if !isSupportedImage(imgBytes) {
+		return preview, fmt.Errorf("unsupported image type url='%s'", u.url.String())
+	}
+
+	compressedBytes, err := compressImage(imgBytes)
+	if err != nil {
+		return preview, fmt.Errorf("failed to compress image url='%s': %w", u.url.String(), err)
+	}
+
+	width, height, err := images.GetImageDimensions(compressedBytes)
+	if err != nil {
+		return preview, fmt.Errorf("could not get image dimensions url='%s': %w", u.url.String(), err)
+	}
+
+	dataURI, err := images.GetPayloadDataURI(compressedBytes)
+	if err != nil {
+		return preview, fmt.Errorf("could not build data URI url='%s': %w", u.url.String(), err)
+	}
+
+	preview.Thumbnail.Width = width
+	preview.Thumbnail.Height = height
+	preview.Thumbnail.DataURI = dataURI
+
+	return preview, nil
+}
+
 func normalizeHostname(hostname string) string {
 	hostname = strings.ToLower(hostname)
 	re := regexp.MustCompile(`^www\.(.*)$`)
 	return re.ReplaceAllString(hostname, "$1")
 }
 
+// isSupportedImageURL detects whether a URL ends with one of the
+// supported image extensions. It provides a quick way to identify whether URLs
+// should be unfurled as images without needing to retrieve the full response
+// body first.
+func isSupportedImageURL(url *neturl.URL) bool {
+	return imageURLRegexp.MatchString(url.Path)
+}
+
+// isSupportedImage returns true when payload is one of the supported image
+// types. In the future, we should differentiate between animated and
+// non-animated WebP because, currently, only static WebP can be processed by
+// functions in the status-go/images package.
+func isSupportedImage(payload []byte) bool {
+	return images.IsJpeg(payload) || images.IsPng(payload) || images.IsWebp(payload)
+}
+
 func newUnfurler(logger *zap.Logger, httpClient http.Client, url *neturl.URL) Unfurler {
+	if isSupportedImageURL(url) {
```

Comment:
Hmmm. This way we won't be able to unfurl image links that don't end with required extension, e.g. https://placekitten.com/100/200. And Discord supports such links.

I think it would be great to decide if it's an image from HTTP header `Content-Type: image/jpeg`. Any chance it's possible to check the header without fetching the whole image?

Reply:
I started coding with this idea in mind @igor-sirotin, but in the end I decided against it in this PR for simplicity sake. To cleanly implement this logic we would need to change the logic that decides which unfurler to create, because the HTTP request would always need to be performed first. And because the request would have been made already, the unfurling logic of all unfurlers would change as well because the response body would already be available.

This should all work fine, unless I face issues related to the user agent. For example, requests to unfurl OpenGraph URLs need to use specific headers, otherwise some websites stubbornly fail. Hopefully the same headers can be used for image URLs.

Since you commented about this limitation, I now have another reason to implement this. I can do it on Monday (I'll be off this Friday).

Reply:
> Any chance it's possible to check the header without fetching the whole image?

If we knew all servers understood HEAD requests, we'd be able to get responses without a body. So I think the best we can do is always request upfront and use that whole information to decide which unfurler to instantiate and pass the response to other functions.

It should work fine because we always fallback to OpenGraph to unfurl, so always requesting upfront won't incur in additional requests, since we would request something anyway.

Reply:
No prob! 
I think it makes sense to skip it for now then and add this later. It will be a status-go update with no breaking changes for desktop/mobile.

Reply:
Perfect!